clippy_lints/src/path_from_format.rs

```diff
@@ -0,0 +1,130 @@
+                let string_parts: Vec<&str> = format_string_parts.iter().map(rustc_span::Symbol::as_str).collect();
+                let mut applicability = Applicability::MachineApplicable;
+                let real_vars: Vec<Sugg<'_>> = format_value_args.iter().map(|x| Sugg::hir_with_applicability(cx, x, "..", &mut applicability)).collect();
+                let order_of_real_vars: Vec<usize> = format_args.formatters.iter().map(|(x, _)| *x).collect();
+                let mut arguments_in_order = Vec::new();
+                for n in 0..real_vars.len() {
+                    arguments_in_order.push(real_vars[order_of_real_vars[n]].clone());
+                }
+                let mut paths_zip = string_parts.iter().take(arguments_in_order.len()).zip(arguments_in_order);
+                let mut sugg = String::new();
+                for n in 0..real_vars.len() {
+                    if let Some((part, arg)) = paths_zip.next() {
+                        if is_valid_use_case(string_parts[n], string_parts.get(n+1).unwrap_or(&"")) {
+                            return;
+                        }
+                        if n == 0 {
+                            if part.is_empty() {
+                                sugg = format!("Path::new({})", arg);
+                            }
+                            else {
+                                push_comps(&mut sugg, part, false);
+                                let _ = write!(sugg, ".join({})", arg);
+                            }
+                        }
+                        else if n < real_vars.len() {
+                            push_comps(&mut sugg, part, true);
+                            let _ = write!(sugg, ".join({})", arg);
+                        }
+                        else {
+                            sugg = format!("{sugg}.join({})", arg);
+                        }
+                    }
+                }
+                if real_vars.len() < string_parts.len() {
+                    push_comps(&mut sugg, string_parts[real_vars.len()], true);
+                }
+                span_lint_and_sugg(
+                    cx,
+                    PATH_FROM_FORMAT,
+                    expr.span,
+                    "`format!(..)` used to form `PathBuf`",
+                    "consider using `Path::new()` and `.join()` to make it OS-agnostic and improve code readability",
+                    sugg,
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        }
+    }
+}
+
+fn push_comps(string: &mut String, path: &str, trim_first_slash: bool) {
+    let mut path = path.to_string();
+    if (path.starts_with('/') || path.starts_with('\\')) && trim_first_slash {
+        path.remove(0);
+    }
+    let path = Path::new(&path);
+    let comps = path.components();
+    for n in comps {
+        let mut x = n.as_os_str().to_string_lossy().to_string();
+        if x.ends_with('/') || x.ends_with('\\') {
+            x.pop();
+        }
```

Comment:
We can use [`str::trim_right_matches`](https://doc.rust-lang.org/std/primitive.str.html#method.trim_end_matches)